app/models/user.rb

```diff
@@ -57,23 +57,33 @@ def destroy
     save!(validate: false)
   end
 
-  def self.from_omniauth(auth) # rubocop:disable Metrics/AbcSize
-    identity = {omniauth_provider: auth.provider, provider_uid: auth.uid}
-
-    user = joins(:identities).where(identities: identity).first_or_initialize do |new_user|
-      # Set these values initially
-      new_user.password = Devise.friendly_token[0, 20]
-      new_user.identities << UserIdentity.new(identity)
-      # If you are using confirmable and the provider(s) you use validate emails,
-      # uncomment the line below to skip the confirmation emails.
-      new_user.skip_confirmation!
+  def self.from_omniauth(auth, user = nil) # rubocop:disable Metrics/AbcSize
+    identity_params = {omniauth_provider: auth.provider, provider_uid: auth.uid}
+    identity = UserIdentity.new(identity_params)
+
+    if user.nil?
+      # A new user signs up with an external account
+      user = joins(:identities).where(identities: identity_params).first_or_initialize do |new_user|
+        # Set these values initially
+        new_user.password = Devise.friendly_token[0, 20]
+        new_user.identities << identity
+        # If you are using confirmable and the provider(s) you use validate emails,
+        # uncomment the line below to skip the confirmation emails.
+        new_user.skip_confirmation!
+      end
+    else
+      # An existing user connects an external account
+      user.identities << identity
     end
 
-    # Update some profile information on every login
-    user.email = auth.info.email
-    user.first_name = auth.info.first_name
-    user.last_name = auth.info.last_name
-    user.save
+    # Update some profile information on every login if present
+    user.assign_attributes(auth.info.slice(:email, :first_name, :last_name).to_h.compact)
```

Comment:
I fixed the issue that prevented users created through OmniAuth to change their profile, especially their avatar.

The update of the said attributes (email, first_name, last_name) is not that trivial and I decided not to touch the behavior in this PR: What do you think a user would expect? Sure, manually updating would be nice. However, if the information (especially the email address) was updated upstream, it would also be nice to store the newest information transferred. Since we get this information on every login, we just assume it's the most current and desired information.

I can think of rarely complex update procedures if we want to change this assumption and behavior, but those are clearly out of scope for this PR.

Reply:
I agree with it being out of scope, but my intuitive approach would be to add a switch, with which the user can control the auto-updating.

Reply:
Alright, I created issue #1329 for that change.